### DISCUSSION TITLE
How to run a docker compose file and run a command in one of the containers

### DISCUSSION THREAD

**Discussion by akkie** [dagger/dagger#4219] [Help]
Hi,

I'm not sure if it's possible in general to run a docker compose file, because I have nothing found in the documentation. Examples are all based on docker files itself.

To the problem. I have a docker compose file which starts different services that I need to execute E2E tests in a pipeline. One of the docker containers that gets started contains my JS based application. The application has a `package.json` which contains one command to run the application and one to run the Playwright E2E tests. I would now like to implement the pipeline with Dagger, but I currently do not have any starting point.

What I need to do is:
- Starting all the required containers based on my docker compose file
- starting the application in one of the containers
- Running the tests

Can someone support me here?

Best regards,
Christian

**Comment by grouville** [comment]
Hi @akkie,

Thanks for this discussion and for trying out Dagger. We'll do out best to help you transition to Dagger smoothly 😇 

By the way, do not hesitate to create a thread on our Discord, under the [help](https://discord.com/channels/707636530424053791/1030538312508776540) section. Also, feel free to ping me (`@guillaume`), or anyone from the team.

To summarize, there is no e2e docs on your use-case because it touches 2 main subjects that are currently under development on our side, as it affects the engine and needs to be carefully designed.

However, it is totally doable, but will require some hacks for the moment.

> We prefer to recommend people to hack their way around the limitations until we have a come up with a satisfying and elegant solution solving most potential follow-up issues on the long-run.

1. **Long running containers**: the ability to run a service on the side, for the entire run of the workflow, against which we run tests in general
2. **Container networking**: how do we expose these containers against each other
3. **Integration with Docker**: you specifically want to run a `docker compose up`, which means that you need a way to interact with the docker daemon, from your Dagger pipeline.

The `3` is already possible. However, the `1` and `2` are still under development, but can be hacked by relying on the docker engine as a long runner, and a container network.
You can find more info on the subject here:
- https://github.com/dagger/dagger/issues/4080
- https://github.com/dagger/dagger/pull/4163/files

> PS: feel free to add your opinion on each of these, as it is currently at the design stage

_Coming back to your question_
1. In order to run a `docker compose` in Dagger, you will need to mount the socket of the docker engine into your dagger pipeline.

We expose 2 APIs for that (taking the Go API as a reference, but works with our Python and Node SDKs too):
* the [host.UnixSocket](https://pkg.go.dev/dagger.io/dagger#Host.UnixSocket), to load your local unix socket as a UnixSocket in Dagger
*  the [container.WithUnixSocket](https://pkg.go.dev/dagger.io/dagger#Container.WithUnixSocket), to load the loaded unixSocket (from the above Host API), into the container in which you will do the `docker compose`.

So, you mount:
- your ` /var/run/docker.sock` into Dagger
- you mount all your code inside the same container
- then, you run the `docker compose up -d`

This will run everything in the local docker daemon.

2. You will need to implement some sort of a healthcheck inside Dagger, to check [when the container gets ready](https://github.com/dagger/dagger/issues/4080#issuecomment-1344580009)

#### Then, not sure to have fully understood this part:
```
starting the application in one of the containers
Running the tests
```
But it seems that you have 2 options (from what I understood):

**Half Dagger / Half Docker**

- You move the app from docker to Dagger (using a dockerfile for example), and it will communicate with the containers created in the local docker daemon
- You also run the tests from Dagger

This will require the containers built from Dagger to be in the same network as the ones created with the above `docker compose up` command.
> But, for this part, feel free to ping us, we can totally help you out. It's not hard at all, but we might help you save time in the process 😇 

**Full Docker**
- As a first step, you just wrap your docker commands inside Dagger. These commands will also run against the local docker daemon (using the socket APIs above).

In that case, Dagger replaces your local bash script. It is a nice first introduction to our API.

### What would I do 
1. I would first try to run the `docker compose up` command against the local docker daemon
2. I would first try the `full docker` option
3. I would then try to slowly move some parts to Dagger (start the app + run the tests)
4. When we finally finish the design of the long running container + container network, you would be able to convert as much as you want from the compose files / this logic to Dagger 😇

**Reply by akkie** [reply]
Thanks for your detailed answer. I think this is a good stating point. I will follow your advice if I need more help.

**Comment by marcosnils** [comment]
Take into account that you'll have to make some modifications to your apps since service resolution by name won't work the same way as it currently works in compose.  More details about this in the issue Guillaume referenced here: 
- https://github.com/dagger/dagger/issues/4080 

So basically, whatever strategy you want to follow today, it'll require some tweaking from the current setup

**Comment by Cirius1792** [comment]
I am interested in this scenario as well. There is any info on when the features 1 and 2 mentioned by @grouville will be implemented?

**Reply by marcosnils** [reply]
@Cirius1792 perfect timing for this. We're releasing our "service containers" (https://docs.dagger.io/757394/use-service-containers) feature this week which addresses this concerns. If you want to know more, @vito did a really good intro with some addittional `compose` sprinkles yesterday in our community call. You can check the recording here: https://dagger-io.zoom.us/rec/share/n6WIx_qMSAU8KiVjm0J-94gUrycpxQWjTwjf6AGXWcvyb_6ztjvfRT_avZoVfduH.KGwTykoBtEreb2oz?startTime=1678381187000

**Reply by Cirius1792** [reply]
It really was good timing! I am already experimenting with the new feature! Thank you!
